### PR TITLE
Mitigate a race condition when package config files are overwritten

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/build/issues/2790

### CODE REVIEW

File: build_runner/test/generate/watch_test.dart [hunk 1]
```diff
@@ -404,9 +404,41 @@ a:file://different/fake/pkg/path
...
+        checkBuild(result, outputs: {'a|web/a.txt.copy': 'a'}, writer: writer);
+
+        await writer.delete(packageConfigId);
+
+        // Wait for it to try reading the file twice to ensure it will retry.
+        await _readerForState[buildState]
+            .onCanRead
+            .where((id) => id == packageConfigId)
+            .take(2)
+            .toList();
```
Comment (natebosch) [inline build_runner/test/generate/watch_test.dart:437]: [nit] 
```suggestion
            .drain();
```

File: build_runner/test/generate/watch_test.dart [hunk 2]
```diff
@@ -816,6 +848,9 @@ Future<BuildState> startWatch(List<BuilderApplication> builders,
       logLevel: logLevel,
       onLog: onLog,
       skipBuildScriptCheck: true);
+  // Some tests need access to `reader` so we expose it through an expando.
```
Comment (natebosch) [inline build_runner/test/generate/watch_test.dart:851]: This feels a little strange to me.

**Review by natebosch — COMMENTED** [review]
Do we expect in the typical case the build will be invalidated? Would it be sensible to catch the error instead and show a nicer error message instead of a stack trace? Will this still have a stack trace if it is missing for a full second?

File: build_runner/test/generate/watch_test.dart (hunk 2 above)
Comment (jakemac53) [reply build_runner/test/generate/watch_test.dart:851]: Ya it is but its just a test and its a low boilerplate/simple way to get ahold of this reader from the test.

I could also create a whole new class which has the reader and BuildState object and returning that (and that is what I would do if this was public code that needed this), but that just seemed overkill for this use case.

**Comment by jakemac53** [comment]
> Do we expect in the typical case the build will be invalidated?

This is specifically to handle the case where somebody runs `pub get/upgrade` while running the server or watcher.

>  Would it be sensible to catch the error instead and show a nicer error message instead of a stack trace?

We could throw a nicer error here, although I do not expect the error case to be hit any more in practice so it didn't seem worth it.

> Will this still have a stack trace if it is missing for a full second?

Not as useful of one, but still useful enough (if somebody were to paste that stack trace with just this one function we could easily still trace what happened). This is never the direct result of user code, its an internal state error essentially.

**Comment by natebosch** [comment]
> This is specifically to handle the case where somebody runs `pub get/upgrade` while running the server or watcher.

Right, I'm wondering whether in the typical case for this situation once we "recover" and read the file after 1 second, we'll throw away the whole build anyway. I can't remember if we restart the whole watcher at that point, or if it exits entirely.

If we expect that most of the time we would hit the case where something changed and we need to bail out, we could do that eagerly instead of waiting a second to see the new file content and decide to do that.

If we think things can continue smoothly in the typical case when this file has changed we can go with this approach.

**Review by natebosch — APPROVED** [review]

**Comment by jakemac53** [comment]
It looks like even in the case of a `pub get` this will terminate, because pub includes the date the file was generated in the package config 🤦‍♂️ . Given that, ya it might just be good to remove all the stuff trying to determine if the package config has actually changed and just exit eagerly.

**Comment by jakemac53** [comment]
@natebosch ok I went ahead and simplified this a bunch, removed the new test and made it bail immediately on any change to these files.

**Review by natebosch — APPROVED** [review]

**Comment by jakemac53** [comment]
Ok, it turns out any call to `pub run` modifies the `.packages` file, causing this to exit. This breaks tests and some real world usage so I reverted the simplification commits and went back to the file content checks and retries.

**Review by natebosch — APPROVED** [review]

**Pull request merged**